### PR TITLE
Do not try to resolve packages that are not declared in dependencies

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
Resolves: https://github.com/embroider-build/embroider/issues/1067

### CODE REVIEW

File: packages/macros/src/babel/dependency-satisfies.ts [hunk 1]
```diff
@@ -12,7 +12,7 @@ export default function dependencySatisfies(path: NodePath<t.CallExpression>, st
   if (path.node.arguments.length !== 2) {
     throw error(path, `dependencySatisfies takes exactly two arguments, you passed ${path.node.arguments.length}`);
   }
-  let [packageName, range] = path.node.arguments;
+  const [packageName, range] = path.node.arguments;
```
Comment (NullVoxPopuli) [inline packages/macros/src/babel/dependency-satisfies.ts:15]: needed to be const due to the function usage in line 35

**Comment by NullVoxPopuli** [comment]
can I make this draft after submission? :upside_down_face: 
I need to:
 - add a test
 - confirm this actually works in my real app

**Review by lifeart — APPROVED** [review]

**Comment by lifeart** [comment]
@NullVoxPopuli I converted it to draft

**Comment by NullVoxPopuli** [comment]
@lifeart thank you!

File: packages/macros/tests/babel/dependency-satisfies.test.ts [hunk 2]
```diff
@@ -123,5 +123,25 @@ describe(`dependencySatisfies`, function () {
       );
       expect(runDefault(code)).toBe(true);
     });
+
+    test('monorepo resolutions resolve correctly', () => {
```
Comment (NullVoxPopuli) [inline packages/macros/tests/babel/dependency-satisfies.test.ts:127]: I _think_ this is pretty close to my setup

File: packages/macros/tests/babel/dependency-satisfies.test.ts [hunk 3]
```diff
@@ -123,5 +123,25 @@ describe(`dependencySatisfies`, function () {
...
+        import { dependencySatisfies } from '@embroider/macros';
+
+        // specified in dependencies
+        export const util = dependencySatisfies('@embroider/util', '*');
+        // not specified as any kind of dep
+        export const webpack = dependencySatisfies('@embroider/webpack', '*');
+      `);
+
+      expect(code).toMatch(/util = true/);
+      expect(code).toMatch(/webpack = false/);
```
Comment (NullVoxPopuli) [inline packages/macros/tests/babel/dependency-satisfies.test.ts:144]: wip
```
    Expected pattern: /util = true/
    Received string:  "// specified in dependencies
    export const util = false; // not specified as any kind of dep·
    export const webpack = false;"
```

**Comment by ef4** [comment]
Some of the existing babel tests were misleading you I think (`process.cwd` has nothing to do with the resolving for example). I cleaned them up.

There is a similar step still needed for the glimmer tests I think.

**Comment by ef4** [comment]
The remaining test failures happen because the PackageCache at this layer doesn't distinguish which Package is the app vs any other library, and `Package.prototype.dependencies` is strict about only listing `devDependencies` if it knows the Package is the app. So at the moment dependencySatisfies will miss devDependencies of apps.

I'm refactoring PackageCache so that threading through the app root path is mandatory, so it can always give correct answers about the app's dependencies. Up until now we hadn't needed that here.

**Comment by NullVoxPopuli** [comment]
Thanks for poking at this! This'll be a huge help!

**Comment by ef4** [comment]
The two remaining failures are just sporadic, I confirmed they're passing locally.

(We have so many different jobs that even low-probability failures in CI are likely to accumulate and cause  a bad run.)

**Comment by ef4** [comment]
This is ready to go. I hijacked your PR and made it a much better change. 😄

One thing we can consider doing next is making packageCache.resolve strict like this by default. There are definitely some cases in embroider/compat where we'd want to opt into non-strict resolve, but it would be a good default.

**Comment by NullVoxPopuli** [comment]
Thank you!! <3

**Comment by anehx** [comment]
Thanks for this fix, this resolves a lot of issues I had with packages using `dependencySatisfies` in a monorepo! :tada: However, since this makes `dependencySatisfies` much stricter ([needing at least a `peerDependency`](https://github.com/emberjs/ember-render-modifiers/pull/58)) we should make sure it's properly documented. Also, this should maybe even be marked as a breaking change? I'm not informed enough on why the `peerDependency` is required to add the documentation myself but I'd really appreciate if someone with that knowledge could add it (@NullVoxPopuli or @ef4 maybe? :wink: ).

Also, I'm wondering if a peer dependency should really be needed - what happens if a package wants to check whether a certain package is installed or not? A peer dependency for such a case would be wrong, no? This might be what @ef4 meant here: https://github.com/embroider-build/embroider/pull/1070#issuecomment-1011435283 ?

**Comment by ef4** [comment]
> what happens if a package wants to check whether a certain package is installed or not? A peer dependency for such a case would be wrong, no? 

For that we use [optional peer dependencies](https://github.com/emberjs/rfcs/blob/master/text/0507-embroider-v2-package-format.md#optional-peer-dependencies).

**Pull request merged**